pandas/core/reshape/merge.py

```diff
@@ -2401,13 +2402,37 @@ def _factorize_keys(
         if not isinstance(lk, BaseMaskedArray) and not (
             # exclude arrow dtypes that would get cast to object
             isinstance(lk.dtype, ArrowDtype)
-            and is_numeric_dtype(lk.dtype.numpy_dtype)
+            and (
+                is_numeric_dtype(lk.dtype.numpy_dtype)
+                or is_string_dtype(lk.dtype)
+                and not sort
+            )
         ):
             lk, _ = lk._values_for_factorize()
 
             # error: Item "ndarray" of "Union[Any, ndarray]" has no attribute
             # "_values_for_factorize"
             rk, _ = rk._values_for_factorize()  # type: ignore[union-attr]
+        elif isinstance(lk.dtype, ArrowDtype) and is_string_dtype(lk.dtype):
+            import pyarrow as pa
+            import pyarrow.compute as pc
+
+            len_lk = len(lk)
+            lk = lk._pa_array  # type: ignore[union-attr, attr-defined]
+            rk = rk._pa_array  # type: ignore[union-attr, attr-defined]
+            dc = pa.concat_arrays(
+                [lk.combine_chunks(), rk.combine_chunks()]
+            ).dictionary_encode()
```

Comment:
I believe `concat_arrays` and `combine_chunks` are both copying here. I think you can avoid both with something like:

```
dc = pc.dictionary_encode(pa.chunked_array(lk.chunks + rk.chunks))
```

Reply:
Yep that's probably true but applying dictionary_encode to a ChunkedArray is awfully slow.

The whole merge op takes 2.2seconds instead of 0.65

Reply:
Yikes! I guess you could avoid one of the copies with:

```
dc = pa.chunked_array(lk.chunks + rk.chunks).combine_chunks().dictionary_encode()
```